external-crates/move/crates/move-compiler/src/linters/constant_naming.rs

```diff
@@ -0,0 +1,95 @@
+    shared::{program_info::TypingProgramInfo, CompilationEnv},
+    typing::{ast as T, visitor::TypingVisitor},
+};
+use move_ir_types::location::Loc;
+use move_symbol_pool::Symbol;
+
+use super::{LinterDiagCategory, LINTER_DEFAULT_DIAG_CODE, LINT_WARNING_PREFIX};
+
+/// Diagnostic information for constant naming violations.
+const CONSTANT_NAMING_DIAG: DiagnosticInfo = custom(
+    LINT_WARNING_PREFIX,
+    Severity::Warning,
+    LinterDiagCategory::ConstantNaming as u8,
+    LINTER_DEFAULT_DIAG_CODE,
+    "Constant name should be in all caps, snake case, pascal case or upper camel case.",
+);
+
+pub struct ConstantNamingVisitor;
+
+impl TypingVisitor for ConstantNamingVisitor {
+    fn visit(
+        &mut self,
+        env: &mut CompilationEnv,
+        _program_info: &TypingProgramInfo,
+        program: &mut T::Program_,
+    ) {
+        for module_def in program
+            .modules
+            .iter()
+            .filter(|(_, _, mdef)| !mdef.attributes.is_test_or_test_only())
+        {
+            env.add_warning_filter_scope(module_def.2.warning_filter.clone());
+            module_def
+                .2
+                .constants
+                .iter()
+                .for_each(|(loc, name, _constant)| {
+                    visit_constant_custom(env, *name, loc);
+                });
+            env.pop_warning_filter_scope();
+        }
+    }
+}
+
+/// Checks if a constant's name adheres to the all caps snake case naming convention.
+fn visit_constant_custom(env: &mut CompilationEnv, name: Symbol, loc: Loc) {
+    if !is_valid_name(name.as_str()) {
+        let uid_msg = format!(
+            "{} should be named using UPPER_CASE_WITH_UNDERSCORES or PascalCase/UpperCamelCase",
+            name.as_str()
+        );
+
+        let diagnostic = diag!(CONSTANT_NAMING_DIAG, (loc, uid_msg));
+        env.add_diag(diagnostic);
+    }
+}
+
+/// Returns `true` if the string is in all caps snake case, including numeric characters.
+fn is_valid_name(name: &str) -> bool {
+    if name
+        .chars()
+        .all(|c| c.is_uppercase() || c == '_' || c.is_numeric())
+        && name.chars().any(char::is_alphabetic)
+    {
+        return true;
+    }
+    // Check for PascalCase/UpperCamelCase
+    // The string must start with an uppercase letter, and only contain alphanumeric characters,
+    // with every new word starting with an uppercase letter.
+    let is_pascal_case = name.chars().enumerate().all(|(i, c)| {
+        if i == 0 {
+            c.is_uppercase()
+        } else {
+            c.is_alphanumeric()
+                && (c.is_lowercase()
+                    || c.is_numeric()
+                    || (c.is_uppercase()
+                        && name
+                            .chars()
+                            .nth(i - 1)
+                            .map(|prev| prev.is_lowercase() || prev.is_numeric())
+                            .unwrap_or(false)))
+        }
+    });
+
+    return is_pascal_case;
```

Comment:
We can't really detect words here, so this should be sufficient 
```suggestion
    let mut chars = name.chars();
    chars.next().unwrap().is_uppercase() &&
    chars.all(|c| c.is_alphanumeric())
```